typing/ctype.ml

```diff
@@ -4924,29 +4935,34 @@ let rec subtype_rec env trace t1 t2 cstrs =
               ~allow_absent:true in
           let cstrs' =
             List.map
-              (fun (n2,t2) -> (trace, List.assoc n2 ntl1, t2, !univar_pairs))
+              (fun (n2,t2) -> cstr trace (List.assoc n2 ntl1) t2)
               ntl2
           in
           if eq_package_path env p1 p2 then cstrs' @ cstrs
           else begin
             (* need to check module subtyping *)
             let snap = Btype.snapshot () in
-            match List.iter (fun (_, t1, t2, _) -> unify env t1 t2) cstrs' with
-            | () when !package_subtype env p1 fl1 p2 fl2 ->
-              Btype.backtrack snap; cstrs' @ cstrs
-            | () | exception Unify _ ->
-              Btype.backtrack snap; raise Not_found
+            let apply_cstr = function
+              | Unify_cstr r -> unify env r.t1 r.t2
+              | Package_error _ -> Btype.backtrack snap; raise Not_found
+            in
+            match List.iter apply_cstr cstrs' with
+            | exception Unify _ -> Btype.backtrack snap; raise Not_found
+            | () ->
+                match !package_subtype env p1 fl1 p2 fl2 with
+                | Ok () -> Btype.backtrack snap; cstrs' @ cstrs
+                | Error x ->
+                    Btype.backtrack snap;
+                    Package_error { trace; error=x }::cstrs
```

Comment:
The unification backtracking is specific to the first-class module path, where we need to unify the `with` constraint before comparing the two first-class module types.

But yes, the `Package_error _` constructor replaces unification constraints that will have failed later because during unification the first-module case calls once again  `package_subtype`, and in both direction.

However, thinking about it, it seems that a better implementation would be move the support of the first class module error to `unify_package_type` and the unification trace. The resulting code would probably be simpler.
I will have a look.

Reply:
The alternative implementation can be found here: https://github.com/ocaml/ocaml/commit/ae8a32b10bcc9fbf512776f84907de04476f9c5d#diff-89663ec75513e9fa325a3ba34185b45f278d35de3491e223bf2c0c49206a2a72 . Normally, I would propose this in a subsequent PR, but this implementation undoes a significant part of the changes from this PR (and is both smaller in code size and more complete in term of coverage). So it might be better to go with this version.